src/python/pants/base/hash_utils.py

```diff
@@ -39,6 +41,49 @@ def hash_file(path, digest=None):
   return digest.hexdigest() if PY3 else digest.hexdigest().decode('utf-8')
 
 
+class CoercingEncoder(json.JSONEncoder):
+  """An encoder which performs coercions in order to serialize many otherwise illegal objects.
+
+  The python documentation (https://docs.python.org/2/library/json.html#json.dumps) states that
+  dict keys are coerced to strings in json.dumps, but this appears to be incorrect -- it throws a
+  TypeError on things we might to throw at it, like a set, or a dict with tuple keys.
+  """
+
+  def _maybe_encode_dict_key(self, key_obj):
+    # If dict keys aren't strings, recursively encode them until they are. Checking for strings here
+    # means we don't touch keys that are already strings (instead of quoting them).
+    if isinstance(key_obj, bytes):
+      # Bytes often occur as dict keys in python 2 code, but in python 3, trying to encode bytes
+      # keys raises a TypeError. We explicitly check for that here and convert to str.
+      return self.default(str(key_obj))
```

Comment:
This conversion won't perform as expected. It will keep the `b` prefix. 
```python
In [1]: x = b'hello'

In [2]: str(x)
Out[2]: "b'hello'"
```
Instead, you have to use `decode('utf-8')`.

Reply:
I had just tagged you in the above comment! Will now address the havoc I've inflicted!